talk-9-js-in-practice-server/slides.md

```diff
@@ -1,29 +1,166 @@
+- Node JS
+- Server-client interaction
+- Async calls vs. Multithreading
+- npm
+- Middleware
+- Routing/Templating
+- MongoDB/Mongoose
+- Other Server Options
+---
+## What is Node?
 - Essentially a headless V8 engine allows for Javascript outside a browser environment
-- Node as a server, Express as the most common and easy Node server
+- Written in C++, very fast/scalable
+- Node itself is not a web server
+- ExpressJS is an implementation of a web server using Node
+
+Note: Not being run in browser, no DOM. Not running on JavaScript itself, C++ running and interpretting JavaScript.
+"node" "console.log('hello');"
+---
+## What is Express?
+- Most popular web application framework for node
+- Helps you manage everything: routes/requests/views
+- Simplifies and speeds up development vs just using Node
+
+Note: Could you create a web server using just Node? Yes. No point, it will require more lines of code, not as secure as using Express.
+"express <appname>" "cd appname" "npm install" "npm start"
+
+---
+## Server-Client Interaction
+- Client sends some request to server
+- Server receives request and sends back a response
+- Brower is an application that allows us to send a request to a server
+- An Express/Node server must be able to accept these requests and send a response
+- "http://my.pitt.edu.com"
+![Client/Server](/http-diagram.jpg?raw=true)
+Note: "http" lets the browser know it wants to communicate with a server using Hypertext Transfer Protocol
+Defines how messages are formmatted and transmitted and what actions Web servers and browers should take in response to various commands.
+"my.pitt.edu.com" is a host name will be looked up in one of many caches (browser, OS, router) then onto DNS servers from ISP for the IP address.
+ 
+
+---
+
+## Asychronous vs Multithreading
+- Asychronous programming allows a unit of work to run separately from the main application
+- Multithreading is just one technique to implement asychronous processes where code can be run by different processors or cores
+- NodeJS is single threaded, instead uses callback functions
+- Callback functions are functions that are passed to another function
 
-### Brief overview of server-client interaction and the web
 
-### Async calls versus multithreading
+---
 
-### npm
+## Example of asychronous programming
 
-### Middleware
+```
+ var result = database.query("SELECT * FROM hugetable");
+ console.log("Hello World");
+```
+vs.
+```
+database.query("SELECT * FROM hugetable", function(rows) {
+       var result = rows;
+    });
+console.log("Hello World");
+```
+Note: In the first example console.log("Hello World"); will not execute until the query returns, if it is a huge database it may take a while.
+In the second example we are passing a callback function which will perform some action once the query is completed, in this case set our result to rows. Console.log() can be ran right away.
+---
+## npm
+- Stands for Node Package Manager
+- Comes packaged with Node
```

Comment:
I think that's only if you've installed via NVM. Otherwise I think you need to install each separately